src/session/session_api.c

```diff
@@ -850,6 +851,78 @@ err:	API_END_RET(session, ret);
 }
 
 /*
+ * __session_transaction_sync --
+ *	WT_SESSION->transaction_sync method.
+ */
+static int
+__session_transaction_sync(WT_SESSION *wt_session, const char *config)
+{
+	WT_CONFIG_ITEM cval;
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_LOG *log;
+	WT_SESSION_IMPL *session;
+	struct timespec end, now;
+	uint64_t wait_secs;
+
+	session = (WT_SESSION_IMPL *)wt_session;
+	conn = S2C(session);
+	/*
+	 * If logging is not enabled there is nothing to do.
+	 */
+	if (!FLD_ISSET(conn->log_flags, WT_CONN_LOG_ENABLED))
+		return (0);
+	SESSION_API_CALL(session, transaction_sync, config, cfg);
+	WT_STAT_FAST_CONN_INCR(session, txn_sync);
+
+	log = conn->log;
+	ret = 0;
+
+	/*
+	 * If there is no background sync LSN in this session, there
+	 * is nothing to do.
+	 */
+	if (WT_IS_INIT_LSN(&session->bg_sync_lsn))
+		goto err;
+
+	/*
+	 * If our LSN is smaller than the current sync LSN then our
+	 * transaction is stable.  We're done.
+	 */
+	if (WT_LOG_CMP(&session->bg_sync_lsn, &log->sync_lsn) <= 0)
+		goto err;
+
+	/*
+	 * Our LSN is not yet stable.  Wait and check again depending on the
+	 * timeout.
+	 */
+	WT_ERR(__wt_config_gets_def(session, cfg, "timeout", 0, &cval));
+	if (cval.len != 0)
+		wait_secs = (uint64_t)cval.val;
+
+	if (wait_secs == 0)
+		WT_ERR(ETIMEDOUT);
```

Comment:
I think what you're missing is that we only get to the point of looking at the wait time if the transaction is known to be not sync'ed.  If this method is called and the transaction is stable we've already returned.  Only if we want to wait do we check the config setting.  So if the `NULL` config gives a default of 0, it means the user wants the answer right away.  I chose the default of 0 so that this defaults to a simple query and immediately returns an answer.  I could certainly be convinced a larger timeout value is a better default.